### PR TITLE
Increase performances by grouping select by entity class

### PR DESCRIPTION
**Pull request by PierrickMartos**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | no
| BC breaks?        | no     
| Related Issue     | /
| Need Doc update   | no

## Describe your change

Actually, the `IndexManager` do a `SELECT` query for each object. This is clearly a drawback in terms of overall performance and kill the advantage of using Algolia.

My small contribution increase performances (if many results, could increase quite a lot) by doing only one query per entity class with a `WHERE IN(ids)` clause.

## What problem is this fixing?

Performance issue.

### CODE REVIEW

File: src/IndexManager.php [hunk 1]
```diff
@@ -151,7 +158,7 @@ public function rawSearch($query, $className, $page = 1, $nbResults = null, arra
             $nbResults = $this->configuration['nbResults'];
         }
 
-        return $this->engine->search($query, $this->getFullIndexName($className), $page, $nbResults, $parameters);
+        return $this->engine->search($query, $this->getFullIndexName($className), $page,  $nbResults, $parameters);
```
Comment (tristanbes) [inline src/IndexManager.php]: extra space here that you don't need :-)

Comment (PierrickMartos) [reply src/IndexManager.php]: Thanks, fix

**Comment by PierrickMartos** [comment]
Hi @tristanbes, do you need something else in order to get it merged? Thanks!

**Comment by tristanbes** [comment]
I'm just a random guy passing by reviewing the PR, not a maintainer :)

**Comment by nunomaduro** [comment]
@PierrickMartos I am going to dive into this next week.

File: src/IndexManager.php [hunk 2]
```diff
@@ -132,12 +133,18 @@ public function search($query, $className, ObjectManager $objectManager, $page =
...
+                $objectsByEntityClass[$entityClass] = [];
+            }
+
+            $objectsByEntityClass[$entityClass][] = $id;
+        }
+
+        foreach ($objectsByEntityClass as $entityClass => $ids) {
             $repo = $objectManager->getRepository($entityClass);
-            $entity = $repo->findOneBy(['id' => $id]);
+            $entities = $repo->findById($ids);
```
Comment (nunomaduro) [inline src/IndexManager.php]: With this approach, we don't loose the *order* of the results?

Comment (PierrickMartos) [reply src/IndexManager.php]: You're right, we loose the order. MySQL return results by id ASC by default. We could use the `ORDER BY FIELD` (https://stackoverflow.com/questions/3799935/mysql-select-where-in-given-order) but not supported by doctrine by default. Any idea @nunomaduro ?

Comment (nunomaduro) [reply src/IndexManager.php]: Yes. Just re-order the objects again some lines after.

Comment (PierrickMartos) [reply src/IndexManager.php]: @nunomaduro done with this commit: https://github.com/algolia/search-bundle/pull/298/commits/77cf8aa57bda935f850133d5a6dbd64e32f0904c

**Comment by ostrolucky** [comment]
What can we do to help merge this? Index performance of this bundle is really bad.

**Comment by PierrickMartos** [comment]
Moved to: https://github.com/algolia/search-bundle/pull/340

**Pull request closed without merging**